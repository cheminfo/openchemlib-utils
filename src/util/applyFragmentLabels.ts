import type { Molecule } from 'openchemlib';
import { SSSearcher } from 'openchemlib';

export interface ApplyFragmentLabelsOptions {
  /**
   * Algorithm to use for substructure searching and matching the fragment to the molecule.
   * overlapping: create list not containing multiple matches sharing exactly the same atoms
   * firstMatch: create matchList with just one match
   * separated: create list of all non-overlapping matches / not optimized for maximum match count
   * rigorous: create list of all possible matches neglecting any symmetries
   * unique: create list of all distinguishable matches considering symmetries
   * @default 'overlapping'
   */
  algorithm?:
    | 'firstMatch'
    | 'separated'
    | 'overlapping'
    | 'rigorous'
    | 'unique';
}

/**
 * Applies fragment labels from a fragment molecule to a target molecule. This method is in
 * place and modifies the target molecule directly.
 * @param molecule - The target molecule to which fragment labels will be applied.
 * @param fragment - The fragment molecule containing the labels to apply.
 * @param options
 * @returns The number of matches found
 */
export function applyFragmentLabels(
  molecule: Molecule,
  fragment: Molecule,
  options: ApplyFragmentLabelsOptions = {},
): number {
  const { algorithm = 'overlapping' } = options;
  const sssearcher = new SSSearcher();
  sssearcher.setMolecule(molecule);
  sssearcher.setFragment(fragment);
  const found = sssearcher.findFragmentInMolecule({ countMode: algorithm });
  const matches = sssearcher.getMatchList();
  for (const match of matches) {
    for (
      let fragmentAtomIndex = 0;
      fragmentAtomIndex < match.length;
      fragmentAtomIndex++
    ) {
      const moleculeAtomIndex = match[fragmentAtomIndex];
      const label = fragment.getAtomCustomLabel(fragmentAtomIndex);
      if (label) {
        molecule.setAtomCustomLabel(moleculeAtomIndex, label);
      }
    }
  }
  return found;
}
