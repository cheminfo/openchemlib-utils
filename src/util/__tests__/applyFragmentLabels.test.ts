import { Molecule } from 'openchemlib';
import { expect, test } from 'vitest';

import { applyFragmentLabels } from '../applyFragmentLabels.ts';

test('default option find all matching', () => {
  const molecule = Molecule.fromSmiles('CCC(=O)CCCCCCC(=O)CC');
  const fragment = Molecule.fromSmiles('CCC(=O)');
  fragment.setAtomCustomLabel(0, 'β');
  fragment.setAtomCustomLabel(1, 'ɑ');
  fragment.setFragment(true);

  const found = applyFragmentLabels(molecule, fragment);

  expect(found).toBe(4);

  const labels = [];
  for (let i = 0; i < molecule.getAllAtoms(); i++) {
    labels.push(molecule.getAtomCustomLabel(i));
  }

  expect(labels).toStrictEqual([
    'β',
    'ɑ',
    null,
    null,
    'ɑ',
    'β',
    null,
    null,
    'β',
    'ɑ',
    null,
    null,
    'ɑ',
    'β',
  ]);
  expect(molecule.toMolfile()).toMatchSnapshot();
});

test('separated, we can not reuse twice the same atoms', () => {
  const molecule = Molecule.fromSmiles('CCC(=O)CCCCCCC(=O)CC');
  const fragment = Molecule.fromSmiles('CCC(=O)');
  fragment.setAtomCustomLabel(0, 'β');
  fragment.setAtomCustomLabel(1, 'ɑ');
  fragment.setFragment(true);

  const found = applyFragmentLabels(molecule, fragment, {
    algorithm: 'separated',
  });

  expect(found).toBe(2);

  const labels = [];
  for (let i = 0; i < molecule.getAllAtoms(); i++) {
    labels.push(molecule.getAtomCustomLabel(i));
  }

  expect(labels).toStrictEqual([
    'β',
    'ɑ',
    null,
    null,
    null,
    null,
    null,
    null,
    'β',
    'ɑ',
    null,
    null,
    null,
    null,
  ]);
  expect(molecule.toMolfile()).toMatchSnapshot();
});
