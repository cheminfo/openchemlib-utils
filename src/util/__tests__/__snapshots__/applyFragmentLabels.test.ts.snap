// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`default option find all matching 1`] = `
"
Actelion Java MolfileCreator 1.0

 14 13  0  0  0  0  0  0  0  0999 V2000
    9.5263    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.6603   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.7942    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.7942    1.0000    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    6.9282   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.0622    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.1962   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.3301    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.4641   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.5981    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.7321   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.7321   -1.5000    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.8660    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  2  0  0  0  0
  3  5  1  0  0  0  0
  5  6  1  0  0  0  0
  6  7  1  0  0  0  0
  7  8  1  0  0  0  0
  8  9  1  0  0  0  0
  9 10  1  0  0  0  0
 10 11  1  0  0  0  0
 11 12  2  0  0  0  0
 11 13  1  0  0  0  0
 13 14  1  0  0  0  0
M  STY  1   1 DAT
M  SLB  1   1   1
M  SAL   1  1   1
M  SDT   1 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   1     9.5263    0.0000    DA    ALL  1       5
M  SED   1 β
M  STY  1   2 DAT
M  SLB  1   2   2
M  SAL   2  1   2
M  SDT   2 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   2     8.6603    0.5000    DA    ALL  1       5
M  SED   2 ɑ
M  STY  1   3 DAT
M  SLB  1   3   3
M  SAL   3  1   5
M  SDT   3 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   3     6.9282    0.5000    DA    ALL  1       5
M  SED   3 ɑ
M  STY  1   4 DAT
M  SLB  1   4   4
M  SAL   4  1   6
M  SDT   4 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   4     6.0622    0.0000    DA    ALL  1       5
M  SED   4 β
M  STY  1   5 DAT
M  SLB  1   5   5
M  SAL   5  1   9
M  SDT   5 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   5     3.4641    0.5000    DA    ALL  1       5
M  SED   5 β
M  STY  1   6 DAT
M  SLB  1   6   6
M  SAL   6  1  10
M  SDT   6 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   6     2.5981    0.0000    DA    ALL  1       5
M  SED   6 ɑ
M  STY  1   7 DAT
M  SLB  1   7   7
M  SAL   7  1  13
M  SDT   7 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   7     0.8660    0.0000    DA    ALL  1       5
M  SED   7 ɑ
M  STY  1   8 DAT
M  SLB  1   8   8
M  SAL   8  1  14
M  SDT   8 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   8     0.0000    0.5000    DA    ALL  1       5
M  SED   8 β
M  END
"
`;

exports[`separated, we can not reuse twice the same atoms 1`] = `
"
Actelion Java MolfileCreator 1.0

 14 13  0  0  0  0  0  0  0  0999 V2000
    9.5263    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.6603   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.7942    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.7942    1.0000    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    6.9282   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.0622    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.1962   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.3301    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.4641   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.5981    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.7321   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.7321   -1.5000    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    0.8660    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  2  0  0  0  0
  3  5  1  0  0  0  0
  5  6  1  0  0  0  0
  6  7  1  0  0  0  0
  7  8  1  0  0  0  0
  8  9  1  0  0  0  0
  9 10  1  0  0  0  0
 10 11  1  0  0  0  0
 11 12  2  0  0  0  0
 11 13  1  0  0  0  0
 13 14  1  0  0  0  0
M  STY  1   1 DAT
M  SLB  1   1   1
M  SAL   1  1   1
M  SDT   1 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   1     9.5263    0.0000    DA    ALL  1       5
M  SED   1 β
M  STY  1   2 DAT
M  SLB  1   2   2
M  SAL   2  1   2
M  SDT   2 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   2     8.6603    0.5000    DA    ALL  1       5
M  SED   2 ɑ
M  STY  1   3 DAT
M  SLB  1   3   3
M  SAL   3  1   9
M  SDT   3 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   3     3.4641    0.5000    DA    ALL  1       5
M  SED   3 β
M  STY  1   4 DAT
M  SLB  1   4   4
M  SAL   4  1  10
M  SDT   4 NOSEARCH_OCL_CUSTOM_LABEL
M  SDD   4     2.5981    0.0000    DA    ALL  1       5
M  SED   4 ɑ
M  END
"
`;
